### DISCUSSION TITLE
Allow specifying assignment targets for primary constructors

### DISCUSSION THREAD

**Discussion by amal-stack** [dotnet/csharplang#7028] [Language Ideas]
At the moment, the scope of the constructor parameters is restricted to the body of the constructor and extended on demand via backing fields as references to them begin to appear somewhere else in the class. This, as many have pointed out, could lead to confusion and the behavior would not maintain parity with that of positional records. 

To be more flexible and explicit, a way to specify the target of the arguments could be implemented. The target can be:  
```csharp
field                     // Assign to a field
get, set                  // Assign to a read-write Property 
get                       // Assign to a read-only Property
readonly field            // Assign to a read-only field
default                   // If current behavior is to be retained
```
The target could be specified in an `as` or `params` clause:
```csharp
class Person(string FirstName, string LastName) as get, set 
{
}
// OR
class Person(string FirstName, string LastName) params : get, set
{
}

```
#### With generics and inheritance:
```csharp

public class Entity(int Id) as get, set {}

public class Person<T> (int id, string firstName, string lastName, T address) : Entity(id)
    as readonly field                            // firstName, lastName, address are readonly fields, Id is a read-write property of base
    where T : class      
{
   
}
```
#### Renaming targets: specify `_` prefixed field names:

```csharp
public class Entity(int Id) params : get, set {}

public class Person<T> (int id, string firstName, string lastName, T address) : Entity(id)
    params : readonly field(_firstName, _lastName, _address)            // rename fields with _ prefix 
    where T : class      
{
   
}
```
`id` is a parameter for the base class. Here: 
```csharp
readonly field(_firstName, _lastName, _address) 
```
The names in parentheses can be used as field names for the `firstName`, `lastName` and `address` parameters respectively.

A named mapping could also be made possible as suggested [here](https://github.com/dotnet/csharplang/discussions/4025#discussioncomment-5089665):
```csharp
readonly field(_firstName: firstName, _lastName: lastName, _address: address) 
```
#### Access modifiers of the target could be specified:
```csharp
// Using params:
params : public get, private set
// Using as:
as private readonly field
```

**Comment by Richiban** [comment]
I think it's a mistake to try to make the primary constructor feature cover all scenarios. After all, in many of the scenarios you've outlined, by the time you've made syntax so flexible that it can cover all scenarios you're left with something that's not much shorter than if you just wrote the members out in the class body.

For example, is 

```cs
public class Person<T> (int id, string firstName, string lastName, T address) : Entity(id)
    params : readonly field(_firstName, _lastName, _address)            // rename fields with _ prefix 
    where T : class      
{
   
}
```

really much better than 

```cs
public class Person<T> (int id, string firstName, string lastName, T address) : Entity(id)
    where T : class      
{
   readonly int _id = id;
   readonly _firstName = firstname;
   readonly string _lastName = lastname;
   readonly T _address = address;
}
```

Although hopefully even this won't be necessary; I'm still holding out hope for readonly locals and parameters (a feature which is given even more importance after primary constructors ship).

**Reply by FaustVX** [reply]
You don't _need_ private fields as primary parameters can be lifted to field
```C#
class C(int i)
{
    public void Inc()
    => i++;

    public override string ToString()
    => i.ToString();
}
```
generates:
```C#
internal class C
{
    [CompilerGenerated]
    private int <i>PC__BackingField; // lifted parameter to field

    public C(int i)
    {
        <i>PC__BackingField = i;
        base..ctor();
    }

    public void Inc()
    {
        <i>PC__BackingField++;
    }

    [System.Runtime.CompilerServices.NullableContext(1)]
    public override string ToString()
    {
        return <i>PC__BackingField.ToString();
    }
}
```
[sharplab.io](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAUEEsC2UECeAwgPYB2yMCcAxjCQsgDQwjVkA+AAgEwAMAWABQvAIzDhAN3wACGjIC8MshADuMogApRPAJQBuYTQB0ASTI1NB4V1EBOTTWtCbPDZqxkYMrLuEBvYRlgmS4AZlCAFhlzSz8hEMUAPh8AalTDFwSQ8JkSSQgEbAATCFDRPhkAFRIAZSpPAHMrIJCFFKxjGvrsMmbnAF8gA)
